# Querying a SPARQL endpoint (knowledge graph) in python
The code below contains everything needed to execute a SPARQL query, manipulate the data it returns, and store the data in a file suitable for rendering with cytoscape.  Your goal is to get this to work locally and attempt to extend it by, for example, altering the query, sorting the results differently, or working with the output in cytoscape to build an informative visualization.  

## Assignment
Hand in your version of this notebook, a short report describing what you attempted to do with it, and a visualization of your results (e.g. as an image exported from cytoscape). 

## Necessary imports

Note you will need to install (e.g. pip install) pandas (which should have come with Anaconda) and SPARQLWrapper on your machine before this will work

In [1]:
import pandas as pd

from pandas.io.json import json_normalize
from SPARQLWrapper import SPARQLWrapper, JSON

## Create query function

The function below takes a SPARQL query string, queries a sparql service, and returns the result as a pandas DataFrame (a table).

In [2]:
def query_wikidata(sparql_query, sparql_service_url):
    """
    Query the endpoint with the given query string and return the results as a pandas Dataframe.
    """
    # create the connection to the endpoint
    # Wikidata enforces now a strict User-Agent policy, we need to specify the agent
    # See here https://www.wikidata.org/wiki/Wikidata:Project_chat/Archive/2019/07#problems_with_query_API
    # https://meta.wikimedia.org/wiki/User-Agent_policy
    sparql = SPARQLWrapper(sparql_service_url, agent="Sparql Wrapper on Jupyter example")  
    
    sparql.setQuery(sparql_query)
    sparql.setReturnFormat(JSON)

    # ask for the result
    result = sparql.query().convert()
    return json_normalize(result["results"]["bindings"])

## Run our SPARQL query
The example here is the drug -> interacts with protein -> encoded by -> gene -> gwas -> disease query pattern. 
Test it here http://tinyurl.com/hwm9388 and explore the results to see you you might adapt it.  Other example queries for the wikidata service:  http://tinyurl.com/j222k6g ,  http://tinyurl.com/gpfr9kj 

In [3]:
sparql_query = """SELECT ?drug ?drugLabel ?gene ?geneLabel ?entrez_id ?disease ?diseaseLabel WHERE {
      ?drug wdt:P129 ?gene_product .   # drug interacts with a gene_product 
      ?gene_product wdt:P702 ?gene .  # gene_product is encoded by a gene
      ?gene wdt:P2293 ?disease .    # gene is genetically associated with a disease 
      ?gene wdt:P351 ?entrez_id .  # get the entrez gene id for the gene 
      # add labels
      SERVICE wikibase:label {
            bd:serviceParam wikibase:language "en" .
      }
    }
    limit 1000
    """
#to query another endpoint, change the URL for the service and the query
sparql_service_url = "https://query.wikidata.org/sparql"
result_table = query_wikidata(sparql_query, sparql_service_url)

## From here on we are using the Python Data Analysis Library "Pandas"
Look for an introduction like this http://synesthesiam.com/posts/an-introduction-to-pandas.html if you get stuck..

Now look at the results of our SPARQL query.  "shape" shows the dimensions (n rows by n cols).  head() shows the column headers (which will be important later) and a sample of the first few rows of data

In [4]:
result_table.shape

(595, 17)

In [5]:
result_table.head()

disease.type                          disease.value diseaseLabel.type  \
0          uri  http://www.wikidata.org/entity/Q12174           literal   
1          uri  http://www.wikidata.org/entity/Q12174           literal   
2          uri  http://www.wikidata.org/entity/Q12174           literal   
3          uri  http://www.wikidata.org/entity/Q12174           literal   
4          uri  http://www.wikidata.org/entity/Q12174           literal   

  diseaseLabel.value diseaseLabel.xml:lang drug.type  \
0            obesity                    en       uri   
1            obesity                    en       uri   
2            obesity                    en       uri   
3            obesity                    en       uri   
4            obesity                    en       uri   

                               drug.value drugLabel.type drugLabel.value  \
0   http://www.wikidata.org/entity/Q60235        literal        Caffeine   
1  http://www.wikidata.org/entity/Q190012        literal       Adenosine   
2  http://www.wikidata.org/entity/Q407308        literal    Theophylline   
3  http://www.wikidata.org/entity/Q729213        literal     Nicardipine   
4  http://www.wikidata.org/entity/Q905783        literal  Istradefylline   

  drugLabel.xml:lang entrez_id.type entrez_id.value gene.type  \
0                 en        literal             140       uri   
1                 en        literal             140       uri   
2                 en        literal             140       uri   
3                 en        literal             140       uri   
4                 en        literal             140       uri   

                                gene.value geneLabel.type  \
0  http://www.wikidata.org/entity/Q4682275        literal   
1  http://www.wikidata.org/entity/Q4682275        literal   
2  http://www.wikidata.org/entity/Q4682275        literal   
3  http://www.wikidata.org/entity/Q4682275        literal   
4  http://www.wikidata.org/entity/Q4682275        literal   

         geneLabel.value geneLabel.xml:lang  
0  adenosine A3 receptor                 en  
1  adenosine A3 receptor                 en  
2  adenosine A3 receptor                 en  
3  adenosine A3 receptor                 en  
4  adenosine A3 receptor                 en

Compare the column names to the first line of the SPARQL query:

SELECT ?drug ?drugLabel ?gene ?geneLabel ?entrez_id ?disease ?diseaseLabel

Each of the SELECTed items (e.g. ?drug) results in 2 columns: one for its data type (either uri or literal) and one for its value (the thing you are probably looking for).  String literal values for labels have an additional column indicating which language they are in.  

For the purposes of this exercise, we can simplify this to focus only on the string names of the drug, disease and gene in each row in the results.  As follows.

## Extract only the columns we care about

In [6]:
simple_table = result_table[["drugLabel.value", "diseaseLabel.value", "geneLabel.value"]]

In [7]:
simple_table.head()

drugLabel.value diseaseLabel.value        geneLabel.value
0        Caffeine            obesity  adenosine A3 receptor
1       Adenosine            obesity  adenosine A3 receptor
2    Theophylline            obesity  adenosine A3 receptor
3     Nicardipine            obesity  adenosine A3 receptor
4  Istradefylline            obesity  adenosine A3 receptor

### Rename the columns of our simple table

Let's delete the "Label.value" portion from the column names.

In [8]:
simple_table = simple_table.rename(columns = lambda col: col.replace("Label.value", ""))

In [9]:
simple_table.head()

drug  disease                   gene
0        Caffeine  obesity  adenosine A3 receptor
1       Adenosine  obesity  adenosine A3 receptor
2    Theophylline  obesity  adenosine A3 receptor
3     Nicardipine  obesity  adenosine A3 receptor
4  Istradefylline  obesity  adenosine A3 receptor

We have just grabbed the three columns we care about and renamed them.

## Count the number of genes per (drug, disease) pair

How many genes link each unique (drug, disease) pair? After counting, order the (drug, disease) pairs in descending order of number of linking genes.  This is one simple method for finding stronger connections between A and C concepts - based on the number of shared B concepts.  

### Make unique (drug, disease) groups and count the number of genes for this group

In [10]:
counts = simple_table.groupby(["drug", "disease"]).size()

In [11]:
counts.head()

drug                                             disease                  
(2S,3S)-2-amino-3-phenylmethoxybutanedioic acid  essential tremor             1
                                                 schizophrenia                1
1,9-Pyrazoloanthrone                             peripheral artery disease    1
2-Aminoethoxydiphenyl borate                     obesity                      1
2-methyl-6-(phenylethynyl)pyridine               Rheumatoid Arthritis         1
dtype: int64

### Turn result into a Pandas data frame (fancy smart table)

In [12]:
counts = counts.to_frame("gene_count")

In [13]:
counts.head()

gene_count
drug                                            disease                              
(2S,3S)-2-amino-3-phenylmethoxybutanedioic acid essential tremor                    1
                                                schizophrenia                       1
1,9-Pyrazoloanthrone                            peripheral artery disease           1
2-Aminoethoxydiphenyl borate                    obesity                             1
2-methyl-6-(phenylethynyl)pyridine              Rheumatoid Arthritis                1

### Put the drug and disease labels back into columns

In [14]:
counts = counts.reset_index()

In [15]:
counts.head()

drug                    disease  \
0  (2S,3S)-2-amino-3-phenylmethoxybutanedioic acid           essential tremor   
1  (2S,3S)-2-amino-3-phenylmethoxybutanedioic acid              schizophrenia   
2                             1,9-Pyrazoloanthrone  peripheral artery disease   
3                     2-Aminoethoxydiphenyl borate                    obesity   
4               2-methyl-6-(phenylethynyl)pyridine       Rheumatoid Arthritis   

   gene_count  
0           1  
1           1  
2           1  
3           1  
4           1

### Sort the table in descending order of genes

In [16]:
counts = counts.sort_values("gene_count", ascending = False)

In [17]:
counts.head()

drug                   disease  gene_count
127        Caffeine                   obesity           3
261      Givinostat                   obesity           2
529  Trichostatin A                   obesity           2
319   Linoleic acid  diabetes mellitus type 2           2
393    Panobinostat                   obesity           2

### Number the results from 0 onwards

In [18]:
counts = counts.reset_index(drop = True)

In [19]:
counts.head()

drug                   disease  gene_count
0        Caffeine                   obesity           3
1      Givinostat                   obesity           2
2  Trichostatin A                   obesity           2
3   Linoleic acid  diabetes mellitus type 2           2
4    Panobinostat                   obesity           2

---

## Add the genes used to link each (drug, disease) pair

Now add another column containing the actual genes linking each (drug, disease) pair.

### Create a dictionary containing the linking genes for each unique pair

In [20]:
linking_genes = dict()
for (drug, disease), small_table in simple_table.groupby(["drug", "disease"]):
    linking_genes[(drug, disease)] = list(small_table["gene"])

### Example: retrieve the linking genes for (caffeine, obesity)

In [21]:
linking_genes[("Caffeine", "obesity")]

['adenosine A3 receptor',
 'inositol 1,4,5-trisphosphate receptor, type 1',
 'RYR2']

### Make a new column containing the linking genes

In [22]:
counts["genes"] = counts[["drug", "disease"]].apply(
    lambda row: linking_genes[(row["drug"], row["disease"])],
    axis = 1
)

In [23]:
counts.head()

drug                   disease  gene_count  \
0        Caffeine                   obesity           3   
1      Givinostat                   obesity           2   
2  Trichostatin A                   obesity           2   
3   Linoleic acid  diabetes mellitus type 2           2   
4    Panobinostat                   obesity           2   

                                               genes  
0  [adenosine A3 receptor, inositol 1,4,5-trispho...  
1     [histone deacetylase 9, histone deacetylase 7]  
2     [histone deacetylase 9, histone deacetylase 7]  
3  [hepatocyte nuclear factor 4, alpha, peroxisom...  
4     [histone deacetylase 9, histone deacetylase 7]

## Save to file

In [24]:
counts.to_csv("drug_disease_count.tsv", sep = '\t', index = False, encoding = 'utf-8')

---

## Make cytoscape file

For the example query, the sorting based on shared gene count seems unsatisfying.  Now, lets look at all the results in a network view to see if any interesting patterns emerge that might shed some light on the data and how we might process it more effectively.  

Below we will create a file suitable for loading into cytoscape.  It will contain the three edge types of interest, linking drugs to genes, diseases to genes, and drugs to diseases e.g.:

source_node	source_type	edge_type	target_node	target_type


### Start with the drug and gene pairs

In [25]:
drug_gene_links = simple_table[["drug", "gene"]]

In [26]:
drug_gene_links.head()

drug                   gene
0        Caffeine  adenosine A3 receptor
1       Adenosine  adenosine A3 receptor
2    Theophylline  adenosine A3 receptor
3     Nicardipine  adenosine A3 receptor
4  Istradefylline  adenosine A3 receptor

### Rename the columns

In [27]:
drug_gene_links = drug_gene_links.rename(columns = {"drug": "source_node", "gene": "target_node"})

In [28]:
drug_gene_links.head()

source_node            target_node
0        Caffeine  adenosine A3 receptor
1       Adenosine  adenosine A3 receptor
2    Theophylline  adenosine A3 receptor
3     Nicardipine  adenosine A3 receptor
4  Istradefylline  adenosine A3 receptor

### Create a new column specifying the source node type

In [29]:
drug_gene_links["source_type"] = "drug"

In [30]:
drug_gene_links.head()

source_node            target_node source_type
0        Caffeine  adenosine A3 receptor        drug
1       Adenosine  adenosine A3 receptor        drug
2    Theophylline  adenosine A3 receptor        drug
3     Nicardipine  adenosine A3 receptor        drug
4  Istradefylline  adenosine A3 receptor        drug

### Create a new column containing the edge type

In [31]:
drug_gene_links["edge_type"] = "interacts_with"

In [32]:
drug_gene_links.head()

source_node            target_node source_type       edge_type
0        Caffeine  adenosine A3 receptor        drug  interacts_with
1       Adenosine  adenosine A3 receptor        drug  interacts_with
2    Theophylline  adenosine A3 receptor        drug  interacts_with
3     Nicardipine  adenosine A3 receptor        drug  interacts_with
4  Istradefylline  adenosine A3 receptor        drug  interacts_with

### Create a new column containing the target type

In [33]:
drug_gene_links["target_type"] = "gene"

In [34]:
drug_gene_links.head()

source_node            target_node source_type       edge_type  \
0        Caffeine  adenosine A3 receptor        drug  interacts_with   
1       Adenosine  adenosine A3 receptor        drug  interacts_with   
2    Theophylline  adenosine A3 receptor        drug  interacts_with   
3     Nicardipine  adenosine A3 receptor        drug  interacts_with   
4  Istradefylline  adenosine A3 receptor        drug  interacts_with   

  target_type  
0        gene  
1        gene  
2        gene  
3        gene  
4        gene

## Repeat for disease gene pairs

In [35]:
disease_gene_links = simple_table[["disease", "gene"]]

In [36]:
disease_gene_links.head()

disease                   gene
0  obesity  adenosine A3 receptor
1  obesity  adenosine A3 receptor
2  obesity  adenosine A3 receptor
3  obesity  adenosine A3 receptor
4  obesity  adenosine A3 receptor

## Rename the columns

In [37]:
disease_gene_links = disease_gene_links.rename(columns = {"disease": "source_node", "gene": "target_node"})

In [38]:
disease_gene_links.head()

source_node            target_node
0     obesity  adenosine A3 receptor
1     obesity  adenosine A3 receptor
2     obesity  adenosine A3 receptor
3     obesity  adenosine A3 receptor
4     obesity  adenosine A3 receptor

### Create the new columns

In [39]:
disease_gene_links["source_type"] = "disease"
disease_gene_links["edge_type"] = "associated_with"
disease_gene_links["target_type"] = "gene"

In [40]:
disease_gene_links.head()

source_node            target_node source_type        edge_type target_type
0     obesity  adenosine A3 receptor     disease  associated_with        gene
1     obesity  adenosine A3 receptor     disease  associated_with        gene
2     obesity  adenosine A3 receptor     disease  associated_with        gene
3     obesity  adenosine A3 receptor     disease  associated_with        gene
4     obesity  adenosine A3 receptor     disease  associated_with        gene

In [41]:
drug_disease_links = (simple_table
    [["drug", "disease"]]
    .assign(
        source_type = "drug",
        edge_type = "may treat",
        target_type = "disease"
    )
    .rename(columns = {"drug": "source_node", "disease": "target_node"})
)

# Join the (disease, gene), (drug, gene), and (drug, disease) tables together

### Number of rows of each table

In [42]:
len(drug_gene_links)

595

In [43]:
len(disease_gene_links)

595

In [44]:
len(drug_disease_links)

595

### Join all three tables together

In [45]:
cytoscape_edges = pd.concat([drug_gene_links, disease_gene_links, drug_disease_links])

In [46]:
cytoscape_edges.head()

edge_type     source_node source_type            target_node  \
0  interacts_with        Caffeine        drug  adenosine A3 receptor   
1  interacts_with       Adenosine        drug  adenosine A3 receptor   
2  interacts_with    Theophylline        drug  adenosine A3 receptor   
3  interacts_with     Nicardipine        drug  adenosine A3 receptor   
4  interacts_with  Istradefylline        drug  adenosine A3 receptor   

  target_type  
0        gene  
1        gene  
2        gene  
3        gene  
4        gene

In [47]:
len(cytoscape_edges)

1785

Note that the final result has 1785 (= 595 * 3) rows. (595 was the original number of results returned)

### Reorder columns

In [48]:
cytoscape_edges = cytoscape_edges[["source_node", "source_type", "edge_type", "target_node", "target_type"]]

In [49]:
cytoscape_edges.head()

source_node source_type       edge_type            target_node  \
0        Caffeine        drug  interacts_with  adenosine A3 receptor   
1       Adenosine        drug  interacts_with  adenosine A3 receptor   
2    Theophylline        drug  interacts_with  adenosine A3 receptor   
3     Nicardipine        drug  interacts_with  adenosine A3 receptor   
4  Istradefylline        drug  interacts_with  adenosine A3 receptor   

  target_type  
0        gene  
1        gene  
2        gene  
3        gene  
4        gene

## Save file to disk

In [50]:
cytoscape_edges.to_csv("drug_gene_disease_network.txt", sep = '\t', index = False, encoding = 'utf-8')

#open a new network in cytoscape and load the file by: File..Import..Network
#use Cytoscape to develop a visualization 
#submit the tab-delimited output files, an image of your network, and an explanation of what you did as your assignment
#Are there any notable hubs in the network?
#Could you extend the code to identify them automatically?
#can you adapt the code to search for a specific drug or a specific disease?